### PR TITLE
RDCC-2390: Delete API - Delete Users by Email Pattern or User ID

### PR DESCRIPTION
**Pull request by AdilOoze**
### JIRA link (if applicable) ###

[RDCC-2390](https://tools.hmcts.net/jira/browse/RDCC-2390)

### Change description ###

New `Delete API` to Delete _(Functional Test)_ Users by `Email Pattern` or `User ID`

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/resources/application.yaml
Comment (AdilOoze) [inline src/main/resources/application.yaml:115]: The changes made here and in `application-functional.yaml` to UP URL will be **reverted** once PR has been approved and QA has been completed successfully.

**Review by kavithagade12 — APPROVED** [review]

File: src/functionalTest/resources/application-functional.yaml
Comment (abhijitdiwan) [inline src/functionalTest/resources/application-functional.yaml:12]: we are pointing to PR

File: src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/util/CaseWorkerReferenceDataClient.java
Comment (abhijitdiwan) [inline src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/util/CaseWorkerReferenceDataClient.java:194]: I think let's move this method in common class Utils. I did that but it may reused accross

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/CaseWorkerRefUsersController.java
Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/CaseWorkerRefUsersController.java:227]: We should hide from swagger and i feel only service should allowed to delete is cwr.
Hence in code we can check if service cwr then only delete else forbidden

**Review by abhijitdiwan — CHANGES_REQUESTED** [review]

File: src/functionalTest/resources/application-functional.yaml
Comment (AdilOoze) [reply src/functionalTest/resources/application-functional.yaml:12]: Please see my comment here: https://github.com/hmcts/rd-caseworker-ref-api/pull/291#discussion_r617643863

We are pointing to PR so that it can use the Delete API in UP which has not yet gone to Master (still in PR)

File: src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/util/CaseWorkerReferenceDataClient.java
Comment (AdilOoze) [reply src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/util/CaseWorkerReferenceDataClient.java:194]: I did not actually create this method, all I have done is format the class...

Comment (abhijitdiwan) [reply src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/util/CaseWorkerReferenceDataClient.java:194]: Yes I know I created this method :) if  possible refactor in util hence we can use across

Comment (AdilOoze) [reply src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/util/CaseWorkerReferenceDataClient.java:194]: Ah okay! Sure I will refactor 😄

**Review by nitishsingla1234 — DISMISSED** [review]

File: src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/util/CaseWorkerReferenceDataClient.java
Comment (SujeethKumarAri) [reply src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/util/CaseWorkerReferenceDataClient.java:194]: I think it is already in util

Comment (AdilOoze) [reply src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/util/CaseWorkerReferenceDataClient.java:194]: @abhijitdiwan  - to echo what @SujeethKumarAri has said, this method is currently in the Util class for Integration Tests

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/CaseWorkerRefUsersController.java
Comment (AdilOoze) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/CaseWorkerRefUsersController.java:228]: After discussion with @shreedhar-hmcts and Sourav, hiding the API is not required, instead we have added a `note` to the API in Swagger, the change was made here: https://github.com/hmcts/rd-caseworker-ref-api/pull/291/commits/a9f6e09846c7ca30fa8784535fc1387afa2a99b3

**Review by nitishsingla1234 — DISMISSED** [review]

File: src/main/resources/application.yaml
Comment (SujeethKumarAri) [reply src/main/resources/application.yaml:115]: Blocking the merge until this being done

**Review by SujeethKumarAri — CHANGES_REQUESTED** [review]
Blocking the merge until we revert the UP endpoint to point to AAT instead of PR

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/CaseWorkerRefUsersController.java
Comment (SujeethKumarAri) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/CaseWorkerRefUsersController.java:243]: Can you do the other way around? like
`"PROD". equalsIgnoreCase(environmentName)` so that we can avoid NPE if  by chance `environmentName ` is null

Comment (SujeethKumarAri) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/CaseWorkerRefUsersController.java:243]: Can you please do the other way around like 
`"PROD". equalsIgnoreCase (environmentName)`? just to avoid NPE if by chance environmentName is null.

Comment (AdilOoze) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/CaseWorkerRefUsersController.java]: Addressed here: https://github.com/hmcts/rd-caseworker-ref-api/pull/291/commits/f6d389a2124d37bc2f29eb85fb6fbc2fb1bdf972

File: src/functionalTest/resources/application-functional.yaml
Comment (SujeethKumarAri) [inline src/functionalTest/resources/application-functional.yaml:12]: should not this be pointed to aat as QA is done?

**Review by SujeethKumarAri — CHANGES_REQUESTED** [review]
I am good with the changes, I shall approve once 
"UP is being pointed to AAT again"

File: src/functionalTest/resources/application-functional.yaml
Comment (AdilOoze) [reply src/functionalTest/resources/application-functional.yaml:12]: Will change this once UP PR is merged

Comment (AdilOoze) [reply src/functionalTest/resources/application-functional.yaml]: Changes have been made here: https://github.com/hmcts/rd-caseworker-ref-api/pull/291/commits/a7946e2fc84d69cefdf576170b18ab77b433bba4

**Review by SujeethKumarAri — APPROVED** [review]

**Review by kavithagade12 — APPROVED** [review]

**Review by shreedhar-hmcts — APPROVED** [review]

**Pull request merged**
